numpy/lib/function_base.py

```diff
@@ -3910,6 +3936,39 @@ def percentile(a,
 
         .. versionchanged:: 1.9.0
             A tuple of axes is supported
+    weights: array_like, optional
+        An array of weights associated with the values in `a`.
+
+        Weights cannot
+
+        * be negative
+        * sum to 0
+
+        However, they can be
+
+        * 0, as long as they do not sum to 0
+        * less than 1.  In this case, all weights are re-normalized by
+            the lowest non-zero weight prior to computation.
```

Comment:
I still think this is wrong and this is a serious problem for me to push it over the finish line.  @lpsinger if you wish to help, maybe you can help clear this up.  If this is wrong, then there needs to be a test that it is fixed as well.

Otherwise, we should look into removing `vectorize`, but I am happy with helping out there and looking into it.

Reply:
@tupui do not feel compelled, but it would be indeed be cool to push this through and getting a bit of scipy stats eyes on it to polish things up will help me push the button :).

Reply:
Thanks I will have a look!

Reply:
I don't understand why we should have something special for weights bellow 1. The formula would work with any positive number (as far as I could see in the literature, weights are only defined as non zero positive numbers and in the end the normalization is done with the sum of weights). It would also not be consistent with `np.average`.

Reply:
@tupui I think some form of "weights" is mentioned in every single chapter of a stats textbook, but they're not all commutable.  If I start from a non-weighted (all weights = 1) case, and picture how quantiles are defined, I actually have a hard time imagining what it means to have a weight that is (0, 1).  In the Hyndman paper, a weight of 1 is a point along an axis.  Then what does a weight of say, 0.1 mean?  Happy to learn.  Thanks.

Reply:
I have not been following all of the subtleties here, but it would be nice to have the same constraints on weights as [numpy.average](https://numpy.org/doc/stable/reference/generated/numpy.average.html):

> The only constraint on weights is that sum(weights) must not be 0.

Reply:
Please disregard my comment above; I see that the existing constraints on the weights are indeed compatible with numpy.average.

Reply:
I have the same reservations as @seberg about different handling of weights depending on their size. I think, a clear definition of a quantile is missing in this discussion based in which tests can be implemented (which are currently missing).

As we have similar discussions for expectiles in scipy, see https://github.com/scipy/scipy/pull/17039#discussion_r1008666577, I'd like to propose one (copied from there), inspired by https://doi.org/10.48550/arXiv.0912.0902:

Statistically speaking, a quantile (at level alpha) is a property `Q` of a distribution function `F`, e.g. `Q(F)` (given some level alpha). Note that a quantile `Q(F)` can be set valued. The empirical version arises by plugging in the empirical distribution `\hat{F}` of a given sample `x`.
In my point of view, sample weights/case weights are defined on this distribution in a relative sense; for integer valued weights, it can be interpreted as repeated observation.
The weighted empirical cdf is `\hat{F}(t) = 1/sum(w) * sum_i w_i 1_{x_i <= t}`.

Implied assumptions and properties:
1. Weights `w_i` are non-negative real numbers.
2. A weight of 2 is equivalent to duplicating this observation.
3. `\hat{F}` as well as `Q(\hat{F})` are invariant under scalar rescaling: `w -> c*w` for `c > 0`
4. The expected value (population level) reads as weighted average on a sample level in the defining equations:
  - Minimisation with strict consistent scoring function `S` for `Q`: `Q(F) = argmin_z E_F(S(z, X))` for all `X~F`
    Sample level: `T(\hat{F}) = argmin_z 1/sum(w) sum_i w_i S(z, x_i)`
  - First order condition / strict identification function `V` for `Q`: `Q(F)` is solution of `E_F(V(z, X)) = 0` for all `X~F`
    Sample level: `T(\hat{F})` is solution of `1/sum(w) sum_i w_i V(z, x_i) = 0`

Notes:
- For quantiles, the solutions are not unique. That's the reason for having all those options for `method` in `numpy.quantile`.
- For quantiles, the most used scoring function is the pinball loss `S(z, x) = (1_{z>=x} - alpha) (z - x)`
- For quantiles, the canonical identification function is `V(z, x) = 1_{z>=x} - alpha`.

I would expect some tests that `E[S]` is minimized or that `E[V] = 0`, with `E` replaced by the sample average.

Reply:
@lorentzenchr can you summarize that into what you think is actionable from NumPy?

The definition in the PR uses a definition where weights are frequency weights not "aweights" or whatever you call it:  The quantile has a constant value around that data point if `weight > 1`.  I am *not* convinced that this is a generally accepted good default.  This definition does also indeed feel weird a bit weird `weights < 1`.

To be honest, unless this clears up, I am tempted to close the PR.  If the weights in the PR are not the right way to go, I would also suggest opening a new PR with adjusted code (and closing this one right away).  The reason is that this PR is getting much too unwieldy anyway with hundreds of comments.

Basically, unless someone we can easily clarify what the correct implementation is, it may be that this is just too advanced statistics to include it in NumPy.
(I don't mind adding weights, but we neither get the best speed here, nor even clearly formulated what the right thing is!)